ism_pci_passthrough: add additional test cases

smitterl opened the pull request:
1. Add test case to cover compatibility with reboot
2. Add test case to cover that ISM is functional

Signed-off-by: Sebastian Mitterle <smitterl@redhat.com>

smitterl commented:
```
JOB ID     : d4049a420b41bfb37af3b3aabdf8d1f41bf7b62e
JOB LOG    : /var/lib/avocado/job-results/job-2023-01-24T05.25-d4049a4/job.log
 (1/1) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.connect: PASS (29.66 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-01-24T05.25-d4049a4/results.html
JOB TIME   : 29.99 s

JOB ID     : 9434dd21c283e5e6438e641c7805441adf303d81
JOB LOG    : /var/lib/avocado/job-results/job-2023-01-24T05.29-9434dd2/job.log
 (1/1) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.happy_path: PASS (16.78 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-01-24T05.29-9434dd2/results.html
JOB TIME   : 17.12 s

JOB ID     : 8ea60dfe8c995b0762efb687b6b13b584624a047
JOB LOG    : /var/lib/avocado/job-results/job-2023-01-24T05.29-8ea60df/job.log
 (1/1) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.reboot: PASS (75.08 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-01-24T05.29-8ea60df/results.html
JOB TIME   : 75.45 s
```

smitterl commented:
```
JOB ID     : 379ac4219d3a4cf8a42f39998627d4c18c4ccd02
JOB LOG    : /var/lib/avocado/job-results/job-2023-01-25T05.35-379ac42/job.log
 (1/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.happy_path: PASS (17.09 s)
 (2/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.reboot: PASS (51.75 s)
 (3/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.reboot_start: PASS (61.08 s)
 (4/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.connect: PASS (29.10 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-01-25T05.35-379ac42/results.html
JOB TIME   : 159.59 s

```

chunfuwen commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -1,10 +1,107 @@
+
+
+def check_guest_and_host_can_communicate(vm, guest_iface):
+    """
```
I assume in general rule, we can use vm.login() to validate communication between guest and host. Is there special reason to use special tool smc_chk here for that?

smitterl replied:
> I assume in general rule, we can use vm.login() to validate communication between guest and host. Is there special reason to use special tool smc_chk here for that?

@chunfuwen Yes, the communication is via the ISM device through the SMC-D protocol. S. https://github.com/autotest/tp-libvirt/blob/a7e7a87de604b2056c49ec5537bcca2c1458d21f/libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py#L62

chunfuwen reviewed (approved):
LGTM

smitterl commented:
@dzhengfy hi, please can you help review?

dzhengfy commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -1,10 +1,107 @@
+    output = session.cmd_output("lspci")
+    session.close()
+    devices = output.split('\n')
+    if not len(devices) >= 1 or "ISM" not in devices[0]:
```
Regarding `"ISM" not in devices[0]`, are you sure there is only 1 device in lspci output? I am curious about it.

smitterl replied:
Yes, this check is inside of the guest so there must be only the ISM device.

dzhengfy commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -1,10 +1,107 @@
+    thread = start_test_server(session)
+    guest_ip = vm.wait_for_get_address(nic_index=0)
+    process.run("systemctl stop firewalld")
+    output = process.run("smc_chk -C %s -p %s" % (guest_ip, smc_port)).stdout_text
```
Do you need to add `verbose=True` to enable the debug info for  process.run()?

smitterl replied:
No, it's the default value - https://github.com/avocado-framework/avocado/blob/master/avocado/utils/process.py#L946

dzhengfy commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -15,27 +112,29 @@ def run(test, params, env):
     finally:
-        if session:
-            session.close()
+        process.run("systemctl stop firewalld", shell=True, verbose=True)
```
This service is already stopped on line 72 and line 57. So there is no need to stop it again, right?

smitterl replied:
Good catch. You are right. The daemon should be started again.

dzhengfy commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -15,27 +112,29 @@ def run(test, params, env):
-            session.close()
+        process.run("systemctl stop firewalld", shell=True, verbose=True)
+        session = vm.wait_for_login()
+        session.cmd("systemctl stop firewalld")
```
same as above

smitterl replied:
You are right. However, the guest will be destroyed so I think it's okay to just omit this line.

dzhengfy reviewed (commented):
Others LGTM

smitterl commented on libvirt/tests/src/passthrough/pci/ism_pci_passthrough.py:
```diff
@@ -1,10 +1,107 @@
 
 
 logging = log.getLogger('avocado.' + __name__)
+smc_port = 37373
```
I don't expect the port to be changed. It's the default port. I declared it here in a variable to avoid the anti-pattern of a magic number (s. 3. in https://www.baeldung.com/cs/antipatterns-magic-numbers). Please, let me know if you prefer it be moved to the cfg file.

smitterl commented:
@dzhengfy Thank you very much for your thorough review. I addressed your comments. Please, let me know if there's anything else.

```
# avocado run --vt-type libvirt --vt-machine-type s390-virtio --vt-connect-uri qemu:///system libvirt_pci_passthrough.ism --test-runner=runner --vt-extra-params "pci_dev = pci_0000_00_00_0"
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 548825f3a92e29dda6b496ca54b28d012af77b3b
JOB LOG    : /var/lib/avocado/job-results/job-2023-02-27T06.41-548825f/job.log
 (1/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.happy_path: PASS (16.03 s)
 (2/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.reboot: PASS (50.74 s)
 (3/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.reboot_start: PASS (61.86 s)
 (4/4) type_specific.io-github-autotest-libvirt.libvirt_pci_passthrough.ism.connect: PASS (27.57 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-02-27T06.41-548825f/results.html
JOB TIME   : 156.65 s
```
